README


This archive contains the distribution POEx-Role-PSGIServer,
version 1.110670:

  Encapsulated PSGI server behaviors for POE

This software is copyright (c) 2011 by Infinity Interactive.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for POEx-Role-PSGIServer: 1.110670

1.110670  2011-03-08 08:02:37 America/Chicago
        Depending on POE::Filter::HTTP::Parser 1.06 because prior versions
        constructed their URI wrong in an edge case that Plack::Test::Suite
        now uncovers.

1.103341  2010-11-30 16:11:52 America/New_York

1.103340  2010-11-30 16:03:34 America/New_York
        Moved the code to K&R brace style to work around a stupid MXD bug
        under 5.12 perls.  Oy ve.

1.103331  2010-11-29 10:18:00 America/New_York
        Declare dependency on latest POEx::Role::TCPServer and use around for
		BUILDARGS rather than method.

1.102531  2010-09-10 13:11:09 America/Los_Angeles
        Throw some defaults on listen_ip and listen_port so we don't explode when plackup doesn't pass them to us (frodwith)

1.102530  2010-09-10 12:14:32 America/Los_Angeles
        Some bugs (poll_cb) in the streaming interface fixed, along with a small test fix (frodwith)

1.101040  2010-04-14 08:35:12 America/Chicago
        BUILDARGS needs to return /all/ of the args, not just the ones Plack cares about

1.101020  2010-04-12 15:39:30 America/Chicago
        Initial Release


lib/POEx/Types/PSGIServer.pm
package POEx::Types::PSGIServer;
BEGIN {
  $POEx::Types::PSGIServer::VERSION = '1.110670';
}

#ABSTRACT: Provides type constraints for use in POEx::Role::PSGIServer
use warnings;
use strict;

use MooseX::Types -declare => [qw/
    PSGIServerContext
    PSGIResponse
    PSGIBody
    HTTPRequest
    HTTPCode
/];

use MooseX::Types::Moose(':all');
use MooseX::Types::Structured(':all');
use POEx::Types(':all');
use HTTP::Request;
use HTTP::Status;
use Plack::Util;
use Scalar::Util;


subtype PSGIServerContext,
    as Dict [
        request => HTTPRequest,
        wheel => Optional[Wheel],
        version => Str,
        protocol => Str, 
        connection => Str,
        keep_alive => Bool,
        chunked => Optional[Maybe[Bool]],
        explicit_length => Optional[Maybe[Bool]],
    ];


subtype HTTPRequest,
    as class_type('HTTP::Request');


subtype HTTPCode,
    as Int,
    where {
        HTTP::Status::is_info($_) 
        || HTTP::Status::is_success($_)
        || HTTP::Status::is_redirect($_)
        || HTTP::Status::is_error($_)
    };


subtype PSGIBody,
    as Ref,
    where {
        Plack::Util::is_real_fh($_)
        || (Scalar::Util::blessed($_) && ($_->can('getline') && $_->can('close')))
    };


subtype PSGIResponse,
    as Tuple [
        HTTPCode,
        ArrayRef,
        Optional[ArrayRef|PSGIBody]
    ];
1;


=pod

=head1 NAME

POEx::Types::PSGIServer - Provides type constraints for use in POEx::Role::PSGIServer

=head1 VERSION

version 1.110670

=head1 TYPES

=head2 PSGIServerContext

PSGIServerContext is defined as a Hash with the following keys:

    request => HTTPRequest,
    wheel => Optional[Wheel],
    version => Str,
    protocol => Str, 
    connection => Str,
    keep_alive => Bool,
    chunked => Optional[Bool],
    explicit_length => Optional[Bool],

The context is passed around to identify the current connection and what it is expecting

=head2 HTTPRequest

This is a simple class_type for HTTP::Request

=head2 HTTPCode

This constraint uses HTTP::Status to check if the Int is a valid HTTP::Status code

=head2 PSGIBody

The PSGIBody constraint covers two of the three types of body responses valid for PGSI responses: a real filehandle or a blessed reference that duck-types getline and close

=head2 PSGIResponse

This constraint checks responses from PSGI applications for a valid HTTPCode, an ArrayRef of headers, and the Optional PSGIBody or ArrayRef body

=head1 AUTHOR

Nicholas Perez <nperez@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Infinity Interactive.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__


lib/POEx/Role/PSGIServer/Streamer.pm
package POEx::Role::PSGIServer::Streamer;
BEGIN {
  $POEx::Role::PSGIServer::Streamer::VERSION = '1.110670';
}

#ABSTRACT: Provides streaming filehandle PSGI implementation
use MooseX::Declare;

class POEx::Role::PSGIServer::Streamer {
    use POE::Filter::Map;
    use POE::Filter::Stream;
    use MooseX::Types::Moose(':all');
    use POEx::Types::PSGIServer(':all');
    

    has server_context => (is => 'ro', isa => PSGIServerContext, required => 1);


    has closed_chunk => ( is => 'rw', isa => Bool, default => 0 );

    with 'POEx::Role::Streaming';


    method _build_filter {
        if($self->server_context->{chunked}) {
            POE::Filter::Map->new(
                Get => sub { $_ },
                Put => sub { 
                    my $data = shift;
                    return $data if $data =~ /0\r\n\r\n/;
                    my $len = sprintf "%X", do { use bytes; length($data) };
                    return "$len\r\n$data\r\n";
                }
            );
        }
        else {
            return POE::Filter::Stream->new();
        }
    }


    around done_writing {
        if($self->server_context->{chunked} && !$self->closed_chunk) {
            $self->closed_chunk(1);
            $self->put("0\r\n\r\n");
            return;
        }

        $self->$orig;
    }
}

1;


=pod

=head1 NAME

POEx::Role::PSGIServer::Streamer - Provides streaming filehandle PSGI implementation

=head1 VERSION

version 1.110670

=head1 PUBLIC_ATTRIBUTES

=head2 server_context

    is: ro, isa: PSGIServerContext, required: 1

This is the server context from POEx::Role::PSGIServer. It is needed to determine the semantics of the current request

=head1 PRIVATE_ATTRIBUTES

=head2 closed_chunk

    is: rw, isa: Bool, default: 0

closed_chunk is a flag used by the advised L</around done_writing> to know whether the chunked transfer encoding needs a terminator or if the terminator has already been written to the output buffer

=head1 PROTECTED_METHODS

=head2 around done_writing

done_writing is advised to check if the context demands a chunked terminator and if one hasn't been sent yet. If so, it marks L</closed_chunk>, puts the terminator into the output buffer, and returns. Upon second invocation when the buffers are flushed, it will execute the original method.

=head1 PRIVATE_METHODS

=head2 _build_filter

_build_filter is overridden to return a L<POE::Filter::Map> filter if the current response is to be chunk transfer encoded. 

=head1 AUTHOR

Nicholas Perez <nperez@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Infinity Interactive.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__


lib/POEx/Role/PSGIServer/ProxyWriter.pm
package POEx::Role::PSGIServer::ProxyWriter;
BEGIN {
  $POEx::Role::PSGIServer::ProxyWriter::VERSION = '1.110670';
}

#ABSTRACT: Provides a push writer for PSGI applications to use
use MooseX::Declare;

class POEx::Role::PSGIServer::ProxyWriter {
    use MooseX::Types::Moose(':all');
    use POEx::Types::PSGIServer(':all');


    has server_context => (
        is => 'ro',
        isa => PSGIServerContext,
        required => 1
    );


    has proxied => (
        is => 'ro',
        isa => Object,
        weak_ref => 1,
        required => 1,
    );


    method write($data) {
        $self->proxied->write($self->server_context, $data);
    }


    method close() {
        $self->proxied->close($self->server_context);
    }


    method poll_cb(CodeRef $coderef) {
        my $on_flush = sub { $self->$coderef() };
        my $id = $self->server_context->{wheel}->ID;
        $self->proxied->set_wheel_flusher($id => $on_flush);
        $on_flush->();
    }
}
1;


=pod

=head1 NAME

POEx::Role::PSGIServer::ProxyWriter - Provides a push writer for PSGI applications to use

=head1 VERSION

version 1.110670

=head1 PUBLIC_ATTRIBUTES

=head2 server_context

    is: ro, isa: PSGIServerContext, required: 1

This is the server context from POEx::Role::PSGIServer. It is needed to determine the semantics of the current request

=head2 proxied

    is: ro, isa: Object, weak_ref: 1, required: 1

This is the actual object that consumes POEx::Role::PSGIServer. It is weakened to make sure it is properly collected when the connection closes

=head1 PUBLIC_METHODS

=head2 write

    ($data)

write proxies to the weakened PSGIServer consumer object passing along the L</server_context>

=head2 close

close is proxied to the weakened PSGIServer consumer passing along L</server_context>

=head2 poll_cb

    (CodeRef $coderef)

poll_cb is provided to complete the interface. The first argument to $coderef will be $self

=head1 AUTHOR

Nicholas Perez <nperez@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Infinity Interactive.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Encapsulated PSGI server behaviors for POE',
  'AUTHOR' => 'Nicholas Perez <nperez@cpan.org>',
  'BUILD_REQUIRES' => {
    'FindBin' => '0',
    'HTTP::Request::Common' => '0',
    'Plack' => '0',
    'Plack::Test::Suite' => '0',
    'Test::More' => '0',
    'Test::Requires' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'POEx-Role-PSGIServer',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'POEx::Role::PSGIServer',
  'PREREQ_PM' => {
    'HTTP::Message::PSGI' => '0',
    'HTTP::Request' => '0',
    'HTTP::Status' => '0',
    'Moose::Autobox' => '0',
    'MooseX::Declare' => '0',
    'MooseX::Types' => '0',
    'MooseX::Types::Moose' => '0',
    'MooseX::Types::Structured' => '0',
    'POE::Filter::HTTP::Parser' => '1.06',
    'POE::Filter::Map' => '0',
    'POE::Filter::Stream' => '0',
    'POEx::Role::Streaming' => '0',
    'POEx::Role::TCPServer' => '1.102740',
    'POEx::Types' => '0',
    'Plack::Util' => '0',
    'Scalar::Util' => '0',
    'aliased' => '0',
    'bytes' => '0'
  },
  'VERSION' => '1.110670',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



